### ISSUE TITLE
[Feature Request]: Early stopping supported by fit

### ISSUE THREAD

**Issue by Jeronymous** [speechbrain/speechbrain#1914]
### 🚀The feature

The validation dataset can be used for model selection, but also for early stopping.
In particular, early stopping prevents from guessing a fixed number of epochs in advance : training stops when the validation loss/error rate decreased for a certain amount of steps.
I think it's a must-have feature (for things like AutoML) that should be straightforward to use.

### Solution outline

There is already `EpochCounterWithStopper`, but passing it to `speechbrain.Brain.fit()` does not do anything special.
It just iterates over the epoch count until the end
```
        # Iterate epochs
        for epoch in epoch_counter:
            self._fit_train(train_set=train_set, epoch=epoch, enable=enable)
            self._fit_valid(valid_set=valid_set, epoch=epoch, enable=enable)
            ...
```
A solution would be to add the method `should_stop(self, current, current_metric)` to the base class `EpochCounter` (always returning False for this class), and use it where appropriate in the `fit()` function.

### Additional context

There was this issue opened a while ago: https://github.com/speechbrain/speechbrain/issues/962

But the link given by @mravanelli is broken
> As far as I remember we already have some recipes using early stopping. You can take a look at this one for instance:
> https://github.com/speechbrain/speechbrain/blob/develop/recipes/ERPCore/P3_decoding/train.py

I looked at the current recipes and did not find "real" early stopping implemented (only early stopping when the loss are not finite).

**Comment by Adel-Moumen** [comment]
Hello @Jeronymous, thank you once again for submitting new feature requests.

Based on your request, it seems like @mravanelli and @TParcollet would be the most appropriate individuals to address your needs.

**Comment by TParcollet** [comment]
@pplantinga and @Gastron might want to have a look as well.
